### ISSUE TITLE
Dimensions of GRU's hidden state mismatch but doesn't throw an error

### ISSUE THREAD

**Issue by benmyara** [pytorch/pytorch#3259]
When the input `h0` to an instance of `nn.GRU` has a wrong shape it doesn’t raise an error (at least in some cases). For instance, if it expects a hidden state of shape `(2, n, p)` but receives one of shape `(4, n, p)` it will not complain. I didn't check with LSTM if the same happens.

**Comment by zou3519** [comment]
@benmyara thank you for the bug report. I've been able to repro it:
```
import torch
from torch.autograd import Variable

input_size = 4
hidden_size = 5
gru = torch.nn.GRU(input_size, hidden_size, 2)
input = Variable(torch.randn(10, 3, input_size))
h0 = Variable(torch.randn(2*1*10000, 3, hidden_size))
gru(input, h0)
```.

#3925 will fix this.

**Comment by soumith** [comment]
Fixed by #3925
